### ISSUE TITLE
warp-tls kills threads

### ISSUE THREAD

**Issue by ndmitchell** [yesodweb/wai#692]
When running Hoogle on a Linux box I track what errors occur when serving up requests. The graph is:

![image](https://user-images.githubusercontent.com/1651197/40860998-4af709c6-65df-11e8-9865-bc80ebce44d7.png)

As you can see, I had no errors for many years, then they started happening about 0-7 times a day since I switched to serving over https, and switching from warp to warp-tls. Given Hoogle gets ~4K hits per day, its a very small percentage, but they all die with:

```
2018-04-07T16:53:20.627584126 49.145.129.207 0.014 /?&hoogle=whileM&scope=&mode=body ERROR: thread killed
```

Thoughts? Is that error harmless? Known?

**Comment by kazu-yamamoto** [comment]
Do you know who prints "thread killed"?
If warp prints, it should be "Thread killed by Warp's timeout reaper".

**Comment by ndmitchell** [comment]
It's coming from my code, specifically: https://github.com/ndmitchell/hoogle/blob/master/src/General/Web.hs#L81-L84

```haskell
        (time,res) <- duration $ try_ $ do s <- act pay; evaluate $ rnf s; return s
        res <- either (fmap Left . showException) (return . Right) res
        logAddEntry log (showSockAddr $ remoteHost req)
            (BS.unpack $ rawPathInfo req <> rawQueryString req) time (either Just (const Nothing) res)
```

Basically I wrap the execution of the operation in `try_` (which is just `try` scoped to `SomeException`) and if that returns `Left` I record the error message. At the point the thread is killed, it's actively trying to respond to a users request.

**Comment by kazu-yamamoto** [comment]
So, while executing `Input -> IO Output`, `ThreadKilled` is received.
Uhhm.

**Comment by kazu-yamamoto** [comment]
Which protocol is used for the error case, HTTP/1.1 or HTTP/2?
HTTPS enables HTTP/2 and old versions of HTTP/2 in Warp is buggy.
Would you tell me the version of Warp?

**Comment by ndmitchell** [comment]
I don't know which version of HTTP is running at the time. Looking, I'm quite a bit behind, so will upgrade and let you know. Current state is:

```
/opt/ghc/8.2.1/lib/ghc-8.2.1/package.conf.d
    Cabal-2.0.0.2
    array-0.5.2.0
    base-4.10.0.0
    binary-0.8.5.1
    bytestring-0.10.8.2
    containers-0.5.10.2
    deepseq-1.4.3.0
    directory-1.3.0.2
    filepath-1.4.1.2
    ghc-8.2.1
    ghc-boot-8.2.1
    ghc-boot-th-8.2.1
    ghc-compact-0.1.0.0
    ghc-prim-0.5.1.0
    ghci-8.2.1
    haskeline-0.7.4.0
    hoopl-3.10.2.2
    hpc-0.6.0.3
    integer-gmp-1.0.1.0
    pretty-1.1.3.3
    process-1.6.1.0
    rts-1.0
    template-haskell-2.12.0.0
    terminfo-0.4.1.0
    time-1.8.0.2
    transformers-0.5.2.0
    unix-2.7.2.2
    xhtml-3000.2.2
/home/www/.ghc/x86_64-linux-8.2.1/package.conf.d
    ListLike-4.5.1
    QuickCheck-2.10.0.1
    aeson-1.2.1.0
    ansi-terminal-0.8.0.1
    ansi-wl-pprint-0.6.8.2
    appar-0.1.4
    asn1-encoding-0.9.5
    asn1-parse-0.9.4
    asn1-types-0.3.2
    async-2.1.1.1
    attoparsec-0.13.2.0
    auto-update-0.1.4
    base-compat-0.9.3
    base-orphans-0.6
    base-prelude-1.2.0.1
    base64-bytestring-1.0.0.1
    blaze-builder-0.4.0.2
    byteable-0.1.1
    byteorder-1.0.4
    bytestring-builder-0.10.8.1.0
    cabal-file-th-0.2.4
    case-insensitive-1.2.0.10
    cereal-0.5.4.0
    clock-0.7.2
    cmdargs-0.10.17
    colour-2.3.4
    conduit-1.2.11
    conduit-1.3.0.1
    conduit-extra-1.1.16
    conduit-extra-1.2.3.2
    conduit-extra-1.3.0
    connection-0.2.8
    cookie-0.4.2.1
    cpphs-1.20.8
    cryptonite-0.24
    data-default-0.7.1.1
    data-default-class-0.1.2.0
    data-default-instances-containers-0.0.1
    data-default-instances-dlist-0.0.1
    data-default-instances-old-locale-0.0.1
    dlist-0.8.0.3
    easy-file-0.2.1
    exceptions-0.8.3
    extra-1.6
    extra-1.6.6
    fast-logger-2.4.10
    fmlist-0.9
    foundation-0.0.13
    generic-deriving-1.11.2
    hashable-1.2.6.1
    haskell-src-exts-1.19.1
    haskell-src-exts-1.20.1
    hourglass-0.2.10
    http-client-0.5.7.0
    http-client-tls-0.3.5.1
    http-conduit-2.2.3.2
    http-conduit-2.3.0
    http-date-0.0.6.1
    http-types-0.9.1
    http2-1.6.3
    integer-logarithms-1.0.2
    iproute-1.7.1
    js-flot-0.8.3
    js-jquery-3.2.1
    lifted-base-0.2.3.11
    memory-0.14.6
    mime-types-0.1.0.7
    mmap-0.5.9
    mmorph-1.1.0
    monad-control-1.0.2.2
    mono-traversable-1.0.8.1
    mtl-2.2.1
    network-2.6.3.2
    network-uri-2.6.1.0
    old-locale-1.0.0.7
    old-time-1.1.0.3
    optparse-applicative-0.14.2.0
    parsec-3.1.11
    pem-0.2.2
    polyparse-1.12
    primitive-0.6.2.0
    process-extras-0.7.2
    psqueues-0.2.3.0
    random-1.1
    resourcet-1.1.9
    resourcet-1.2.0
    scientific-0.3.5.1
    simple-sendfile-0.2.25
    socks-0.5.5
    split-0.2.3.3
    stm-2.4.4.1
    storable-record-0.0.3.1
    storable-tuple-0.0.3.3
    streaming-commons-0.1.18
    stringsearch-0.3.6.6
    syb-0.7
    tagged-0.8.5
    tar-0.5.0.3
    text-1.2.2.2
    tf-random-0.5
    time-locale-compat-0.1.1.3
    tls-1.4.0
    tls-session-manager-0.0.0.1
    transformers-base-0.4.4
    transformers-compat-0.5.1.4
    typed-process-0.2.1.0
    uniplate-1.6.12
    unix-compat-0.4.3.1
    unix-time-0.3.7
    unliftio-core-0.1.1.0
    unordered-containers-0.2.8.0
    utf8-string-1.0.1.1
    utility-ht-0.0.14
    uuid-types-1.0.3
    vault-0.3.0.7
    vector-0.12.0.1
    vector-algorithms-0.7.0.1
    void-0.7.2
    wai-3.2.1.1
    wai-extra-3.0.22.0
    wai-logger-2.3.0
    warp-3.2.13
    warp-tls-3.2.4
    word8-0.1.3
    x509-1.7.2
    x509-store-1.6.5
    x509-system-1.6.6
    x509-validation-1.6.9
    zlib-0.6.1.2
```

**Comment by ndmitchell** [comment]
Upgraded to as per below. Will let you know if that fixes it in a few days.

```
/opt/ghc/8.2.1/lib/ghc-8.2.1/package.conf.d
    Cabal-2.0.0.2
    array-0.5.2.0
    base-4.10.0.0
    binary-0.8.5.1
    bytestring-0.10.8.2
    containers-0.5.10.2
    deepseq-1.4.3.0
    directory-1.3.0.2
    filepath-1.4.1.2
    ghc-8.2.1
    ghc-boot-8.2.1
    ghc-boot-th-8.2.1
    ghc-compact-0.1.0.0
    ghc-prim-0.5.1.0
    ghci-8.2.1
    haskeline-0.7.4.0
    hoopl-3.10.2.2
    hpc-0.6.0.3
    integer-gmp-1.0.1.0
    pretty-1.1.3.3
    process-1.6.1.0
    rts-1.0
    template-haskell-2.12.0.0
    terminfo-0.4.1.0
    time-1.8.0.2
    transformers-0.5.2.0
    unix-2.7.2.2
    xhtml-3000.2.2
/home/www/.ghc/x86_64-linux-8.2.1/package.conf.d
    ListLike-4.6
    QuickCheck-2.11.3
    aeson-1.3.1.1
    appar-0.1.4
    asn1-encoding-0.9.5
    asn1-parse-0.9.4
    asn1-types-0.3.2
    async-2.2.1
    attoparsec-0.13.2.2
    auto-update-0.1.4
    base-compat-0.10.1
    base-orphans-0.7
    basement-0.0.7
    blaze-builder-0.4.1.0
    bsb-http-chunked-0.0.0.2
    byteable-0.1.1
    byteorder-1.0.4
    bytestring-builder-0.10.8.1.0
    case-insensitive-1.2.0.11
    cereal-0.5.5.0
    clock-0.7.2
    cmdargs-0.10.20
    conduit-1.3.0.2
    conduit-extra-1.3.0
    connection-0.2.8
    cookie-0.4.4
    cpphs-1.20.8
    cryptonite-0.25
    data-default-0.7.1.1
    data-default-class-0.1.2.0
    data-default-instances-containers-0.0.1
    data-default-instances-dlist-0.0.1
    data-default-instances-old-locale-0.0.1
    dlist-0.8.0.4
    easy-file-0.2.2
    exceptions-0.10.0
    extra-1.6.8
    fast-logger-2.4.11
    fmlist-0.9.2
    foundation-0.0.20
    generic-deriving-1.12.1
    hashable-1.2.7.0
    haskell-src-exts-1.20.2
    hourglass-0.2.11
    http-client-0.5.12.1
    http-client-tls-0.3.5.3
    http-conduit-2.3.1
    http-date-0.0.7
    http-types-0.12.1
    http2-1.6.3
    integer-logarithms-1.0.2.1
    iproute-1.7.5
    js-flot-0.8.3
    js-jquery-3.3.1
    memory-0.14.16
    mime-types-0.1.0.7
    mmap-0.5.9
    mono-traversable-1.0.8.1
    mtl-2.2.2
    network-2.7.0.0
    network-uri-2.6.1.0
    old-locale-1.0.0.7
    old-time-1.1.0.3
    parsec-3.1.13.0
    pem-0.2.4
    polyparse-1.12
    primitive-0.6.4.0
    process-extras-0.7.4
    psqueues-0.2.7.0
    random-1.1
    resourcet-1.2.1
    scientific-0.3.6.2
    semigroups-0.18.4
    simple-sendfile-0.2.27
    socks-0.5.6
    split-0.2.3.3
    stm-2.4.5.0
    storable-record-0.0.4
    storable-tuple-0.0.3.3
    streaming-commons-0.2.0.0
    syb-0.7
    tagged-0.8.5
    tar-0.5.1.0
    text-1.2.3.0
    tf-random-0.5
    th-abstraction-0.2.6.0
    time-locale-compat-0.1.1.4
    tls-1.4.1
    tls-session-manager-0.0.0.2
    transformers-compat-0.6.2
    typed-process-0.2.2.0
    uniplate-1.6.12
    unix-compat-0.5.0.1
    unix-time-0.3.8
    unliftio-core-0.1.1.0
    unordered-containers-0.2.9.0
    utf8-string-1.0.1.1
    utility-ht-0.0.14
    uuid-types-1.0.3
    vault-0.3.1.1
    vector-0.12.0.1
    vector-algorithms-0.7.0.1
    wai-3.2.1.2
    wai-logger-2.3.2
    warp-3.2.22
    warp-tls-3.2.4.3
    word8-0.1.3
    x509-1.7.3
    x509-store-1.6.6
    x509-system-1.6.6
    x509-validation-1.6.10
    zlib-0.6.2
```

**Comment by ndmitchell** [comment]
It's definitely not fixed. I still get ~10 errors a day out of ~3000 uses.

**Comment by larskuhtz** [comment]
I wonder if the following is related: We have an event source, that when served via HTTP works fine, but when served via HTTPS terminates after about 60s, but sometimes also earlier. On the server side the application receives `ThreadKilled`. On the network side the TLS connection is closed normally with `Alert (Level: Warning, Description: Close Notify)` and the TCP connection is closed.

This happens with browsers and curl as clients. It does not happen with `http-client` as client. It was tested with warp-tls-3.2.9.

**Comment by kazu-yamamoto** [comment]
This should be fixed with Warp 3.3.6.
Please reopen if necessary.
